### PR TITLE
Add secondary, cmake based build system

### PR DESCRIPTION
**Pull request by junghans**
## Purpose

Provide an alternative build system for LAMMPS.

## Author(s)

@junghans @rbberger

## Backward Compatibility

Cannot build everything, yet, but I am working on it. However, I should not conflict with existing build system.

## Implementation Notes

It compiles the same files, there should ultimately no differences in the correctness.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [x] The feature or features in this pull request is complete
   -  [x] Packages
      - [x] ASPHERE
      - [x] BODY
      - [x] CLASS2
      - [x] COLLOID
      - [x] COMPRESS
      - [x] CORESHELL
      - [x] DIPOLE
      - [x] GPU
        - [x] SINGLE_DOUBLE SINGLE_SINGLE DOUBLE_DOUBLE
      - [x] GRANULAR
      - [x] KIM
      - [x] KOKKOS
      - [x] KSPACE
        - [x] FFTs
          - [x] FFTW3
          - [x] MKL
          - [x] FFTW2
        - [x] Pack options
      - [x] MANYBODY
      - [x] MC
      - [x] MEAM
      - [x] MISC
      - [x] MOLECULE
      - [x] MPIIO
      - [x] MSCG
      - [x] OPT
      - [x] PERI
      - [x] POEMS
      - [x] PYTHON
      - [x] QEQ
      - [x] REAX
      - [x] REPLICA
      - [x] RIGID
      - [x] SHOCK
      - [x] SNAP
      - [x] SRD
      - [x] USER-ATC
      - [x] USER-AWPMD
      - [x] USER-CGDNA
      - [x] USER-CGSDK
      - [x] USER-COLVARS
      - [x] USER-DIFFRACTION
      - [x] USER-DPD
      - [x] USER-DRUDE
      - [x] USER-EFF
      - [x] USER-FEP
      - [x] USER-H5MD
      - [x] USER-INTEL
      - [x] USER-LB
      - [x] USER-MANIFOLD
      - [x] USER-MEAMC
      - [x] USER-MGPT
      - [x] USER-MISC
      - [x] USER-MOLFILE
      - [x] USER-NETCDF
      - [x] USER-OMP
      - [x] USER-PHONON
      - [x] USER-QMMM
      - [x] USER-QTB
      - [x] USER-QUIP
      - [x] USER-REAXC
      - [x] USER-SMD
      - [x] USER-SMTBQ
      - [x] USER-SPH
      - [x] USER-TALLY
      - [x] USER-VTK
      - [x] VORONOI
    - [x] External libs/programs
      - [x] PNG
      - [x] GZIP
      - [x] FFMPEG
      - [x] JPEG
      - [x] Lapack
        - [x] internal  Lapack
        - [x] external Lapack (ATLAS, mkl,...) through cmake
    - [x] Global Defines 
      - [x] SMALL_SMALL, BIG_BIG, BIG_SMALL
      - [x] LAMMPS_MEMALIGN
      - [x] LAMMPS_XDR
      - [x] LAMMPS_EXCEPTIONS
    - [x] Different compilers (gcc, icc, clang) 
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] One or more example input decks are included
- [x] The source code follows the LAMMPS formatting guidelines

## Test this feature ##
```
git fetch origin pull/573/head
git checkout FETCH_HEAD
mkdir build
cd build
cmake ../cmake
make -j5
./lmp
```

Feedback please.....

### CODE REVIEW

**Comment by rbberger** [comment]
@junghans been there before and was making the case as well. I even got the whole style header generation figured out. But it gets ugly when you reach certain packages. Here is an old version of my [first attempt](https://github.com/rbberger/lammps/blob/cmake_and_gtest/src/CMakeLists.txt). My current variant is [here](https://github.com/rbberger/lammps/blob/cmake_build/src/CMakeLists.txt) which I started after this year's MD workshop at Temple. Note there are also CMakeList.txt files in the package folders. Maybe you'll get inspired by it. But I don't see this going anywhere with @sjplimp and @akohlmey. Maybe we can make the case of adding it as secondary way of compiling it...

**Comment by junghans** [comment]
@rbberger: I like rbberger/lammps@e6711079081666b6756c0227bf6906afa6956e76, but I would

- not put `CMakeLists.txt` into src and the package directories to not conflict with the existing system (like I did in the pull request)
- in `RegisterStyles()` one could add a directory option and hand it down to the `GLOB` in `FindStyleHeaders()`, so could performe the whole lammps build in one CMakeLists.txt, without a single `add_subdirectory()`.
- Plus, we could move that single `CMakeLists.txt` into a subfolder `cmake/` to hide it from the common user.

With these changes, it seems reasonable to me to push this as an _alternative_ build system (like I said above) and see if users like it. Before that I won't even port documentation and testing to cmake. 

One the plus side:

- native Windows support (no shell needed, support for MS compiler and MS nmake)
- `cmake` has `ccmake` and `cmake-gui`, which are nice GUIs for beginners to use
- no copying around of sources anymore
- build could be done with ninja or even Eclipse instead of gmake.
- ultimately that would also fix #400 as cmake plays well with [rpm](https://fedoraproject.org/wiki/Packaging:Cmake).

**Comment by rbberger** [comment]
One of the problems I was running into were packages which install files dependent on what was already installed. E.g., like OMP variants of only the existing styles. So one has to enforce a certain order and keep information about those styles in-memory to allow making these decisions in subsequent package inclusions.

**Comment by junghans** [comment]
I see, the cmake way™ would anyway be to not call `RegisterStyles()`, but create `RegisterStyles.cmake` script to be called aat make time (not cmake time) and simply define a dependency rules like:
```cmake
add_custom_command(OUTPUT style_atom.h COMMAND ${CMAKE_COMMAND} -P ${CMAKE_MODULE_PATH}/RegisterStyple.cmake DEPEND <*_atom*>)
```
and let cmake do its magic.

**Comment by junghans** [comment]
@sjplimp @akohlmey : Before I invest more time into cmake, are there any major objections to provide a secondary buildsystem?

**Comment by akohlmey** [comment]
@junghans i suggest you convene a little "cmake summit" at the upcoming LAMMPS workshop where you and @rbberger can discuss with @sjplimp. It will help if you can demonstrate some of the basic principles, e.g. how you can have toggles in the GUI for setting -D flags (e.g. for JPEG/PNG/FFT) and more.
I would **love** to have a system in place that allows "out-of-source" builds. However, there are many complications (GPU support, KOKKOS, Fortran, python interface, external libraries, OpenKIM), it will be a significant effort to make it work for everything.

My major requirement for adding this would be a promise to maintain and support it (including helping people on the mailing list) in a timely fashion.

**Comment by junghans** [comment]
@akohlmey: I added

- support for external libjpeg
- support for one package as a test
- out-source build work
- replace use of `make.sh` by using pre-generated `style_*.h` with conditionals in them. These could be auto-generate by a script and only needed to be touched in case a package would be added anyhow.

**Comment by junghans** [comment]
@akohlmey :`REAX` and hence Fortran support works now, too.

**Comment by rbberger** [comment]
@junghans The `add_custom_command` and DEPENDS part doesn't make the problems go away. They're the wrong tool for the job. It's the wrong kind of "dependency" you are looking at. We already "know" everything at configuration time (aka. CMake time).

**Comment by rbberger** [comment]
@junghans It's probably best if we split up packages into ones which have a simple installation (all it's contents) and detect all the styles and then do the difficult ones like the accelerator packages, which do a selection based on all the packages enabled so far.

I have an idea how to proceed and would rewrite my RegisterStyles to work within one CMakeLists.txt. But it would replace your hard-coded style headers. Again, the arguments against those is having to maintain them and adding a burden to simple extensions.

**Comment by rbberger** [comment]
@junghans added a package checklist to this PR description so we can keep track of what we have

**Comment by junghans** [comment]
@rbberger: `OPT` package needs your help, I will look at `USER` packages next. `KIM` should be done in same spirit as `VORONOID` (if there is a `kim-api-build-config` use it).

**Comment by junghans** [comment]
@rbberger: how about testing?

**Comment by rbberger** [comment]
@junghans I have a lot of other things going on up to the LAMMPS workshop. I'll try to squeeze in some testing of USER-INTEL, GPU and KOKKOS. If you're talking about Jenkins integration of the build process, that is something I'll look into once it is merged to master.

**Comment by junghans** [comment]
@rbberger: no worries, I was mainly interested in running the jenkins testsuite through `ctest` (aka`make test`).

**Comment by junghans** [comment]
Ready for review.

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+GenerateStyleHeaders(${LAMMPS_STYLE_HEADERS_DIR})
+
+include_directories(${LAMMPS_SOURCE_DIR})
+include_directories(${LAMMPS_STYLE_HEADERS_DIR})
+
+
+add_library(lammps ${LIB_SOURCES})
+if(ENABLE_GPU)
+  add_dependencies(lammps cubin_headers)
```
Comment (junghans) [inline cmake/CMakeLists.txt]: @chuckatkins: Is there are better way to trigger the header generation?

File: lib/gpu/lal_pppm_d.cu [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#ifdef CMAKE_GPU
+#define grdtyp double
```
Comment (junghans) [inline lib/gpu/lal_pppm_d.cu]: @chuckatkins: Are there per-source compiler defines for cuda targets?

**Comment by junghans** [comment]
@sjplimp: Can you have a look at the listing above to see if we are missing any major feature?

**Comment by sjplimp** [comment]
looks complete to me ... thx

**Comment by junghans** [comment]
1. We didn't add OpenCL support, even though it doesn't seem hard: https://streamhpc.com/blog/2015-09-25/handling-opencl-with-cmake-3-1-and-higher/, we should add this to #616.
2. `lal_pppm.cu` gets compiled twice in `Nvidia.makefile`, with different compiler defines (`-Dgrdtyp=float -Dgrdtyp4=float4` and `-Dgrdtyp=double -Dgrdtyp4=double4`). And CMake-3.1 has no support for per-cuda-object compiler defines, yet, so the two `cu` files outside of `lib/gpu` just set the same define and include `lal_pppm.cu`.

**Comment by sjplimp** [comment]
I'm fine with releasing this before all possible LAMMPS build
options are enabled for CMake.  However, there needs to
be some documentation added to doc/src/Section_start.txt
to indicate how this can be used.  I think it should have enough
content that users unfamiliar with CMake can try it out.

**Comment by akohlmey** [comment]
@sjplimp this is not meant to be a release. given the complexity of the task and the fact, that a major milestone has been reached, richard and i have both approved the PR to be merged. there will be another PR based on the action items listed in issue #616, and that one **will** include documentation and only that is supposed to be used by regular users. 
merging this PR now will not affect any other feature (and cmake support should not be announced), yet it will significantly ease further development.

**Pull request merged**